test/module/irohad/ametsuchi/postgres_query_executor_test.cpp

```diff
@@ -1188,44 +1207,205 @@ namespace iroha {
       shared_model::crypto::Hash fake_hash{zero_string};
       shared_model::crypto::PublicKey fake_pubkey{zero_string};
-      const std::string asset_id = "coin#domain";
-      std::unique_ptr<shared_model::interface::Account> account2;
       shared_model::crypto::Hash hash1;
       shared_model::crypto::Hash hash2;
       shared_model::crypto::Hash hash3;
     };
 
-    class GetAccountTransactionsExecutorTest
+
+    template <typename QueryTxPaginationTest>
+    class GetPagedTransactionsExecutorTest
         : public GetTransactionsExecutorTest {
-     public:
+     protected:
+      using Impl = QueryTxPaginationTest;
+
       // create valid transactions and commit them
-      std::vector<shared_model::proto::Transaction> createTransactionsAndCommit(
-          size_t size,
-          const shared_model::interface::types::AccountIdType &creator) {
-        std::vector<shared_model::proto::Transaction> txs;
-        txs.reserve(size);
-        for (size_t i = 0; i < size; i++) {
-          txs.push_back(TestTransactionBuilder()
-                            .creatorAccountId(account->accountId())
-                            .createRole("user" + std::to_string(i), {})
-                            .build());
+      void createTransactionsAndCommit(size_t transactions_amount) {
+        addPerms(Impl::getUserPermissions());
+
+        auto initial_txs = Impl::makeInitialTransactions(transactions_amount);
+        auto target_txs = Impl::makeTargetTransactions(transactions_amount);
+
+        tx_hashes_.reserve(target_txs.size());
+        initial_txs.reserve(initial_txs.size() + target_txs.size());
+        for (auto &tx : target_txs) {
+          tx_hashes_.emplace_back(tx.hash());
+          initial_txs.emplace_back(std::move(tx));
         }
 
         auto block = TestBlockBuilder()
-                         .transactions(txs)
+                         .transactions(initial_txs)
                          .height(1)
                          .prevHash(fake_hash)
                          .build();
 
         apply(storage, block);
+      }
+
+      auto queryPage(
+          types::TransactionsNumberType page_size,
+          const boost::optional<types::HashType> &first_hash = boost::none) {
+        auto query = Impl::makeQuery(page_size, first_hash);
+        return executeQuery(query);
+      }
+
+      /**
+       * Check the transactions pagination response compliance to general rules:
+       * - total transactions number is equal to the number of target
+       * transactions
+       * - the number of transactions in response is equal to the requested
+       * amount if there are enough, otherwie equal to the available amount
+       * - the returned transactions' and the target transactions' hashes match
+       * - next transaction hash in response is unset if the last transaction is
+       * in the response, otherwise it matches the next target transaction hash
+       */
+      void GeneralTransactionsPageResponseCheck(
```

Comment:
This function contains several checks. Maybe move these checks to separate functions, so that this one contains calls to them. This will make it easier to read and understand where each check starts.

Reply:
These checks reuse several common intermediate results, such as the relation to target tx hashes. And I see no use case for separate functions. So I would leave it as is while it is not so big.